### ISSUE TITLE
EzXML vs LightXML

### ISSUE THREAD

**Issue by henry2004y** [henry2004y/Vlasiator.jl#29]
Here are some benchmarks for the two XML parsers:

```julia
using LightXML

function foo(filename, varname)
   fid = open(filename, "r")
   # First 8 bytes indicate big-endian or else
   endian_offset = 8
   seek(fid, endian_offset)
   # Obtain the offset of the XML file
   offset = read(fid, UInt64)
   seek(fid, offset)
   footer = read(fid, String) |> parse_string |> root

   arraysize, datasize, datatype, vectorsize, variable_offset = 0, 0, "", 0, 0
   isFound = false

   for varinfo in footer["VARIABLE"]
      if attribute(varinfo, "name") == String(varname)
         arraysize = parse(Int, attribute(varinfo, "arraysize"))
         datasize = parse(Int, attribute(varinfo, "datasize"))
         datatype = attribute(varinfo, "datatype")
         vectorsize = parse(Int, attribute(varinfo, "vectorsize"))
         variable_offset = parse(Int, content(varinfo))
         isFound = true
         break
      end
   end

   close(fid)

   return variable_offset, arraysize, datasize, vectorsize
end


filename = "bulk.0000000.vlsv"

@time foo(filename, "proton/vg_rho")
```

```julia
# Load the package.
using EzXML

function foo2(filename, name)
   fid = open(filename, "r")
   # First 8 bytes indicate big-endian or else
   endian_offset = 8
   seek(fid, endian_offset)
   # Obtain the offset of the XML file
   offset = read(fid, UInt64)
   seek(fid, offset)
   footer = read(fid, String) |> parsexml |> root

   arraysize, datasize, datatype, vectorsize, variable_offset = 0, 0, "", 0, 0
   isFound = false

   vars = findall("//VARIABLE", footer)
   for var in vars
      if var["name"] == String(name)
         arraysize = parse(Int, var["arraysize"])
         datasize = parse(Int, var["datasize"])
         datatype = var["datatype"]
         vectorsize = parse(Int, var["vectorsize"])
         variable_offset = parse(Int, nodecontent(var))
         isFound = true
         break
      end
   end

   close(fid)

   return variable_offset, arraysize, datasize, vectorsize
end


filename = "bulk.0000000.vlsv"

@time foo2(filename, "proton/vg_rho")
```

LightXML:
```julia
julia> include("test_lightxml.jl") # 1st execution
  0.084423 seconds (117.03 k allocations: 7.140 MiB, 9.33% gc time, 99.62% compilation time)
(124784, 312, 8, 1)

julia> @time foo(filename, "proton/vg_rho")
  0.000295 seconds (180 allocations: 18.047 KiB)
(124784, 312, 8, 1)

julia> @btime foo(filename, "proton/vg_rho")
  102.765 μs (180 allocations: 18.05 KiB)
(124784, 312, 8, 1)
```

EzXML:
```julia
julia> include("test_ezxml.jl") # 1st execution
  0.274732 seconds (723.06 k allocations: 42.738 MiB, 2.76% gc time, 99.89% compilation time)
(124784, 312, 8, 1)

julia> @time foo2(filename, "proton/vg_rho")
  0.000387 seconds (39 allocations: 8.516 KiB)
(124784, 312, 8, 1)

julia> @btime foo2(filename, "proton/vg_rho")
  96.936 μs (39 allocations: 8.52 KiB)
(124784, 312, 8, 1)
```

1. LightXML is faster to load.
2. EzXML consumes less memory.
3. On average ExXML is faster.

**Comment by henry2004y** [comment]
With preset benchmarks:
Current v0.7.9 with LightXML
```julia
| ID                         | time            | GC time | memory          | allocations |
|----------------------------|----------------:|--------:|----------------:|------------:|
| `["VTK", "AMR"]`           |  37.208 ms (5%) |         |  11.37 MiB (1%) |      289302 |
| `["read", "DCCRG"]`        |  29.521 μs (5%) |         | 133.81 KiB (1%) |         178 |
| `["read", "DCCRG_select"]` |  14.529 ms (5%) |         |   1.12 MiB (1%) |       36151 |
| `["read", "FG"]`           | 372.201 μs (5%) |         | 437.69 KiB (1%) |        4528 |
| `["read", "VDF"]`          | 339.971 μs (5%) |         |   1.51 MiB (1%) |         631 |
| `["read", "meta"]`         | 500.545 μs (5%) |         | 399.42 KiB (1%) |        3752 |
```
Replaced with EzXML:
```julia
| ID                         | time            | GC time | memory          | allocations |
|----------------------------|----------------:|--------:|----------------:|------------:|
| `["VTK", "AMR"]`           |  42.302 ms (5%) |         |  11.28 MiB (1%) |      287824 |
| `["read", "DCCRG"]`        |  35.741 μs (5%) |         | 124.12 KiB (1%) |          27 |
| `["read", "DCCRG_select"]` |  14.703 ms (5%) |         |   1.10 MiB (1%) |       35849 |
| `["read", "FG"]`           | 390.752 μs (5%) |         | 408.84 KiB (1%) |        4077 |
| `["read", "VDF"]`          | 374.052 μs (5%) |         |   1.47 MiB (1%) |          69 |
| `["read", "meta"]`         | 522.623 μs (5%) |         | 176.56 KiB (1%) |         304 |
```

**Comment by henry2004y** [comment]
I really have some doubts in the PkgBenchmark test suites. In any case, I create a simple script and run BenchmarkTools directly.
* 1st timing is about reading metadata;
* 2nd timing is about reading `proton/vg_rho` in sorted order
* 3rd timing is about reading `proton/vg_rho` in unsorted order

TL;DR:
* EzXML takes longer to compile for the 1st time execution.
* The speed difference is almost negligble.
* EzXML has lower memory consumption.

## filename: bulk.singleprecision.vlsv 9KB, 1D
rho: 80 KB single precision
* LightXML
  181.754 μs (2882 allocations: 191.50 KiB)
  11.647 μs (154 allocations: 9.69 KiB)
  10.818 μs (153 allocations: 9.53 KiB)

* EzXML
  213.775 μs (282 allocations: 23.64 KiB)
  16.908 μs (16 allocations: 880 bytes)
  16.465 μs (15 allocations: 720 bytes)

## filename: bulk.0000003.vlsv 47MB, 3D
rho: 900 KB double precision
* LightXML
  7.073 ms (3983 allocations: 2.84 MiB)
  351.965 μs (181 allocations: 2.16 MiB)
  98.877 μs (175 allocations: 889.23 KiB)

* EzXML
  6.020 ms (310 allocations: 2.60 MiB)
  355.870 μs (27 allocations: 2.15 MiB)
  90.089 μs (21 allocations: 879.38 KiB)

## filename: bulk1.1001000.vlsv 26GB, 3D
rho: 32 MB double precision
* LightXML
  344.456 ms (7946 allocations: 91.30 MiB)
  11.409 ms (255 allocations: 30.28 MiB)
  2.540 ms (253 allocations: 15.15 MiB)
* EzXML
  322.711 ms (355 allocations: 90.82 MiB)
  11.269 ms (31 allocations: 30.26 MiB)
  2.530 ms (29 allocations: 15.13 MiB)

## filename: bulk.2d.vlsv 900KB, 2D
rho: 50 KB double precision
* LightXML
  537.143 μs (3752 allocations: 399.42 KiB)
  27.091 μs (178 allocations: 133.81 KiB)
  16.158 μs (172 allocations: 59.73 KiB)
* EzXML
  521.827 μs (304 allocations: 176.56 KiB)
  33.262 μs (27 allocations: 124.12 KiB)
  22.347 μs (21 allocations: 50.05 KiB)

**Comment by henry2004y** [comment]
As a side note, I list some numbers using Analysator. Note that Analysator 0.9 does not perform sorting in reading meta data, so a meaningful comparison should be made by adding the two numbers (meta load + var read) together.

## filename: bulk.singleprecision.vlsv
read meta : 2139 [μs]
read DCCRG: 101  [μs]

## filename: bulk.0000003.vlsv
read meta : 2314 [μs]
read DCCRG: 7703 [μs]

## filename: bulk1.0001000.vlsv
read meta : 2455 [μs]
read DCCRG: 292 [ms]
